onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -260,20 +269,52 @@ bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
     return true;
   }
 
+  bool are_input_same = inputs_.size() == other.inputs_.size();
+  
+  if(are_input_same) {
+    for(unsigned long i=0; i<inputs_.size(); i++) {
+      std::vector<const EquivalenceClass*> iclasses_ = inputs_[i];
+      std::vector<const EquivalenceClass*> oclasses_ = other.inputs_[i];
+      
+      are_input_same = are_input_same && (iclasses_.size() == oclasses_.size());
+      
+      if(!are_input_same)
+        break;
+      
+      for(unsigned long j=0; j<iclasses_.size(); j++) {
+        const EquivalenceClass* clsi = iclasses_[j];
+        const EquivalenceClass* clso = oclasses_[j];
+        
+        if((clsi->initializer_ != nullptr) && (clso->initializer_ != nullptr)) {
+          are_input_same = are_input_same && (clsi->initializer_->raw_data()
+                                              == clso->initializer_->raw_data());
+        } else {
+          are_input_same = are_input_same && (clsi->non_op_value_ == clso->non_op_value_);
+        }
+      }
+    }
+  }
+
   // Below we compare inputs_ as pointers. This is valid due to how EquivalenceClass'es are constructed:
   // we'll never have two distinct but equal inputs_ here, so their addresses are effectively their value numbers.
   return hash_ == other.hash_ && output_index_ == other.output_index_ && discriminator_ == other.discriminator_ &&
-         non_op_value_ == other.non_op_value_ &&
+         //non_op_value_ == other.non_op_value_ &&
          op_type_ == other.op_type_ && domain_ == other.domain_ &&
-         inputs_ == other.inputs_ &&
+         are_input_same &&
          SameAttributes(attributes_, other.attributes_);
 }
 
 std::size_t EquivalenceClass::CalculateHash() const {
   std::size_t hash = 0;
   UpdateHash(output_index_, hash);
   UpdateHash(discriminator_, hash);
-  UpdateHash(non_op_value_, hash);
+  
+  if(initializer_ != nullptr) {
+    UpdateHash(initializer_->raw_data(), hash);
```

Comment:
I think comparing just `raw_data` is insufficient. Check documentation for `TensorProto` [here](https://github.com/onnx/onnx/blob/master/onnx/onnx.proto), you need to compare at least the shape (`dims`), and probably `float_data`, `int32_data` etc...

Reply:
Also, I don't know if something special needs to be done in training. @yuslepukhin, are trained parameters represented as inputs (not initializers) during training?